compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -708,6 +708,41 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
     else text
   }
 
+  /** Whether the counterexample is satisfiable. The space is flattened and non-empty. */
+  def satisfiable(sp: Space): Boolean = {
+    def genConstraint(space: Space): List[(Type, Type)] = space match {
+      case Prod(tp, unappTp, unappSym, ss, _) =>
+        val tps = signature(unappTp, unappSym, ss.length)
+        ss.zip(tps).flatMap {
+          case (sp : Prod, tp) => sp.tp -> tp :: genConstraint(sp)
+          case (Typ(tp1, _), tp2) => tp1 -> tp2 :: Nil
+          case _ => ???  // impossible
```

Comment:
I feel that we should use something other than `???` for unreachable cases, to me `???` really has the semantic of a TODO: "implementation is missing here, to be added later"

Maybe something like this? Not sure where should that definition could go...

```scala
def impossible: Nothing = throw new AssertionError("This expression shouldn't be reachable.")
```

Reply:
`assert(false, "Unreachable code")`?

Reply:
@allanrenucci assert has type Unit

Reply:
My bad...